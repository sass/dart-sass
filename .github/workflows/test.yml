name: Test

defaults:
  run: {shell: bash}

on:
  workflow_call:
  workflow_dispatch:

jobs:
  format:
    name: Code formatting
    runs-on: ubuntu-latest

    steps:
    - uses: actions/checkout@v4
    - uses: dart-lang/setup-dart@v1
    - run: dart format --fix .
    - run: git diff --exit-code

  static_analysis:
    name: Static analysis
    runs-on: ubuntu-latest

    steps:
      - uses: actions/checkout@v4
      - uses: ./.github/util/initialize
        with: {github-token: "${{ github.token }}"}

      - name: Analyze Dart
        run: dart analyze --fatal-warnings ./

  dartdoc:
    name: Dartdoc
    runs-on: ubuntu-latest

    steps:
      - uses: actions/checkout@v4
      - uses: ./.github/util/initialize
        with: {github-token: "${{ github.token }}"}

      - name: dartdoc sass
        run: dart run dartdoc --quiet --no-generate-docs
          --errors ambiguous-doc-reference,broken-link,deprecated
          --errors unknown-directive,unknown-macro,unresolved-doc-reference
      - name: dartdoc sass_api
        run: cd pkg/sass_api && dart run dartdoc --quiet --no-generate-docs
          --errors ambiguous-doc-reference,broken-link,deprecated
          --errors unknown-directive,unknown-macro,unresolved-doc-reference

  sass_spec_language:
    name: "Language Tests | Dart ${{ matrix.dart_channel }} | ${{ matrix.async_label }}"
    runs-on: ubuntu-latest

    strategy:
      fail-fast: false
      matrix:
        dart_channel: [stable, dev]
        async_label: [synchronous]
        async_args: ['']
        include:
          - dart_channel: stable
            async_label: asynchronous
            async_args: '--cmd-args --async'

    steps:
      - uses: actions/checkout@v4
      - uses: ./.github/util/initialize
        with:
          dart-sdk: ${{ matrix.dart_channel }}
          github-token: ${{ github.token }}
      - uses: ./.github/util/sass-spec

      - name: Run specs
        run: npm run sass-spec -- --dart .. $extra_args
        working-directory: sass-spec
        env: {extra_args: "${{ matrix.async_args }}"}

  sass_spec_js:
    name: "JS API Tests | Pure JS | Dart ${{ matrix.dart_channel }} | Node ${{ matrix.node-version }} | ${{ matrix.os }}"
    runs-on: "${{ matrix.os }}"

    strategy:
      fail-fast: false
      matrix:
        os: [ubuntu-latest, windows-latest, macos-latest]
        dart_channel: [stable]
        node-version: ['lts/*']
        include:
          # Test older LTS versions
          - os: ubuntu-latest
            dart_channel: stable
            node-version: lts/-1
          - os: ubuntu-latest
            dart_channel: stable
            node-version: lts/-2
          - os: ubuntu-latest
            dart_channel: stable
            node-version: lts/-3
          # Test LTS version with dart dev channel
          - os: ubuntu-latest
            dart_channel: dev
            node-version: 'lts/*'

    steps:
      - uses: actions/checkout@v4
      - uses: ./.github/util/initialize
        with:
          dart-sdk: ${{ matrix.dart_channel }}
          github-token: ${{ github.token }}
          node-version: ${{ matrix.node-version }}
      - uses: ./.github/util/sass-spec

      - name: Build JS
        run: dart run grinder pkg-npm-dev

      - name: Check out Sass specification
        uses: sass/clone-linked-repo@v1
        with:
          repo: sass/sass
          path: language

      - name: Run tests
        run: npm run js-api-spec -- --sassSassRepo ../language --sassPackage ../build/npm
        working-directory: sass-spec

  sass_spec_js_embedded:
    name: 'JS API Tests | Embedded | Node ${{ matrix.node-version }} | ${{ matrix.os }}'
    runs-on: ${{ matrix.os }}
    if: "github.event_name != 'pull_request' || !contains(github.event.pull_request.body, 'skip sass-embedded')"

    strategy:
      fail-fast: false
      matrix:
        os: [ubuntu-latest, windows-latest, macos-latest]
        node-version: ['lts/*']
        include:
          # Test older LTS versions
          - os: ubuntu-latest
            dart_channel: stable
            node-version: lts/-1
          - os: ubuntu-latest
            dart_channel: stable
            node-version: lts/-2
          - os: ubuntu-latest
            dart_channel: stable
            node-version: lts/-3

    steps:
      - uses: actions/checkout@v4
      - uses: ./.github/util/initialize
        with:
          github-token: ${{ github.token }}
          node-version: ${{ matrix.node-version }}
      - uses: ./.github/util/sass-spec

      - name: Check out the embedded host
        uses: sass/clone-linked-repo@v1
        with: {repo: sass/embedded-host-node}

      - name: Check out the language repo
        uses: sass/clone-linked-repo@v1
        with: {repo: sass/sass, path: build/language}

      - name: Initialize embedded host
        run: |
          npm install
          npm run init -- --compiler-path=.. --language-path=../build/language
          npm run compile
          mv {`pwd`/,dist/}lib/src/vendor/dart-sass
        working-directory: embedded-host-node

      - name: Version info
        run: |
          path=embedded-host-node/dist/lib/src/vendor/dart-sass/sass
          if [[ -f "$path.cmd" ]]; then "./$path.cmd" --version
          elif [[ -f "$path.bat" ]]; then "./$path.bat" --version
          elif [[ -f "$path.exe" ]]; then "./$path.exe" --version
          else "./$path" --version
          fi

      - name: Run tests
        run: npm run js-api-spec -- --sassPackage ../embedded-host-node --sassSassRepo ../build/language
        working-directory: sass-spec

  sass_spec_js_browser:
    name: "JS API Tests | Browser | Dart ${{ matrix.dart_channel }}"

    strategy:
      matrix:
        dart_channel: [stable]
      fail-fast: false

    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v4
      - uses: browser-actions/setup-chrome@v1
      - uses: ./.github/util/initialize
        with:
          dart-sdk: ${{ matrix.dart_channel }}
          github-token: ${{ github.token }}
      - uses: ./.github/util/sass-spec

      - name: Build JS
        run: dart run grinder pkg-npm-dev

      - name: Install built dependencies
        run: npm install
        working-directory: build/npm

      - name: Check out Sass specification
        uses: sass/clone-linked-repo@v1
        with:
          repo: sass/sass
          path: language

      - name: Run tests
        run: npm run js-api-spec -- --sassSassRepo ../language --sassPackage ../build/npm --browser
        working-directory: sass-spec
        env:
          CHROME_EXECUTABLE: chrome

  dart_tests:
    name: "Dart tests | Dart ${{ matrix.dart_channel }} | ${{ matrix.os }}"
    runs-on: "${{ matrix.os }}"

    strategy:
      fail-fast: false
      matrix:
        os: [ubuntu-latest, windows-latest, macos-latest]
        dart_channel: [stable]
        # TODO(nweiz): Re-enable this when
        # https://github.com/dart-lang/sdk/issues/52121#issuecomment-1728534228
        # is addressed.
        # include: [{os: ubuntu-latest, dart_channel: dev}]

    steps:
      - uses: actions/checkout@v4
      - uses: ./.github/util/initialize
        with:
          dart-sdk: ${{ matrix.dart_channel }}
          github-token: ${{ github.token }}

      - run: dart run grinder pkg-standalone-dev
      - name: Run tests
        run: dart run test -x node

  # Unit tests that use Node.js, defined in test/.
  node_tests:
    name: "Node tests | Dart ${{ matrix.dart_channel }} | Node ${{ matrix.node-version }} | ${{ matrix.os }}"
    runs-on: "${{ matrix.os }}"

    strategy:
      fail-fast: false

      matrix:
        os: [ubuntu-latest, windows-latest, macos-latest]
        dart_channel: [stable]
        node-version: ['lts/*']
        include:
          # Test older LTS versions
          - os: ubuntu-latest
            dart_channel: stable
            node-version: lts/-1
          - os: ubuntu-latest
            dart_channel: stable
            node-version: lts/-2
          - os: ubuntu-latest
            dart_channel: stable
            node-version: lts/-3
          # Test LTS version with dart dev channel
          - os: ubuntu-latest
            dart_channel: dev
            node-version: 'lts/*'

    steps:
      - uses: actions/checkout@v4
      - uses: ./.github/util/initialize
        with:
          dart-sdk: ${{ matrix.dart_channel }}
          github-token: ${{ github.token }}
          node-version: ${{ matrix.node-version }}

      - run: dart run grinder pkg-npm-dev
      - name: Run tests
        run: dart run test -t node -j 2

  browser_tests:
    name: "Browser Tests | Dart ${{ matrix.dart_channel }}"

    strategy:
      matrix:
        dart_channel: [stable]
      fail-fast: false

    runs-on: ubuntu-latest
    steps:
    - uses: actions/checkout@v4
    - uses: browser-actions/setup-chrome@v1
    - uses: ./.github/util/initialize
      with:
        dart-sdk: ${{ matrix.dart_channel }}
        github-token: ${{ github.token }}

    - run: dart run grinder pkg-npm-dev
    - name: Run tests
      run: dart run test -p chrome -j 2
      env:
        CHROME_EXECUTABLE: chrome
