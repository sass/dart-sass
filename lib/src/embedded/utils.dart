// Copyright 2019 Google Inc. Use of this source code is governed by an
// MIT-style license that can be found in the LICENSE file or at
// https://opensource.org/licenses/MIT.

import 'dart:typed_data';

import 'package:protobuf/protobuf.dart';
import 'package:source_span/source_span.dart';
import 'package:term_glyph/term_glyph.dart' as term_glyph;

import '../syntax.dart';
import 'embedded_sass.pb.dart' as proto;
import 'embedded_sass.pb.dart' hide SourceSpan, Syntax;
import 'util/varint_builder.dart';

/// The special ID that indicates an error that's not associated with a specific
/// inbound request ID.
const errorId = 0xffffffff;

/// Returns a [ProtocolError] indicating that a mandatory field with the given
/// [fieldName] was missing.
ProtocolError mandatoryError(String fieldName) =>
    paramsError("Missing mandatory field $fieldName");

/// Returns a [ProtocolError] indicating that the parameters for an inbound
/// message were invalid.
ProtocolError paramsError(String message) => ProtocolError()
  // Set the ID to [errorId] by default. This will be overwritten by the
  // dispatcher if a request ID is available.
  ..id = errorId
  ..type = ProtocolErrorType.PARAMS
  ..message = message;

/// Returns a [ProtocolError] with type `PARSE` and the given [message].
ProtocolError parseError(String message) => ProtocolError()
  ..type = ProtocolErrorType.PARSE
  ..message = message;

/// Converts a Dart source span to a protocol buffer source span.
proto.SourceSpan protofySpan(SourceSpan span) {
  var protoSpan = proto.SourceSpan()
    ..text = span.text
    ..start = _protofyLocation(span.start)
    ..end = _protofyLocation(span.end)
    ..url = span.sourceUrl?.toString() ?? "";
  if (span is SourceSpanWithContext) protoSpan.context = span.context;
  return protoSpan;
}

/// Converts a Dart source location to a protocol buffer source location.
SourceSpan_SourceLocation _protofyLocation(SourceLocation location) =>
    SourceSpan_SourceLocation()
      ..offset = location.offset
      ..line = location.line
      ..column = location.column;

/// Converts a protocol buffer syntax enum into a Sass API syntax enum.
Syntax syntaxToSyntax(proto.Syntax syntax) {
  switch (syntax) {
    case proto.Syntax.SCSS:
      return Syntax.scss;
    case proto.Syntax.INDENTED:
      return Syntax.sass;
    case proto.Syntax.CSS:
      return Syntax.css;
    default:
      throw "Unknown syntax $syntax.";
  }
}

/// Returns the result of running [callback] with the global ASCII config set
/// to [ascii].
T withGlyphs<T>(T callback(), {required bool ascii}) {
  var currentConfig = term_glyph.ascii;
  term_glyph.ascii = ascii;
  var result = callback();
  term_glyph.ascii = currentConfig;
  return result;
}

/// Serializes [value] to an unsigned varint.
Uint8List serializeVarint(int value) {
  if (value == 0) return Uint8List.fromList([0]);
  RangeError.checkNotNegative(value);

  // Essentially `(value.bitLength / 7).ceil()`, but without getting floats
  // involved.
  var lengthInBytes = (value.bitLength + 6) ~/ 7;
  var list = Uint8List(lengthInBytes);
  for (var i = 0; i < lengthInBytes; i++) {
    // The highest-order bit indicates whether more bytes are necessary to fully
    // express the number. The lower 7 bits indicate the number's value.
    list[i] = (value > 0x7f ? 0x80 : 0) | (value & 0x7f);
    value >>= 7;
  }
  return list;
}

/// Serializes a compilation ID and protobuf message into a packet buffer as
/// specified in the embedded protocol.
Uint8List serializePacket(int compilationId, GeneratedMessage message) {
  var varint = serializeVarint(compilationId);
  var protobufWriter = CodedBufferWriter();
  message.writeToCodedBufferWriter(protobufWriter);

  var packet = Uint8List(varint.length + protobufWriter.lengthInBytes);
  packet.setAll(0, varint);
  protobufWriter.writeTo(packet, varint.length);
  return packet;
}

/// A [VarintBuilder] that's shared across invocations of [parsePacket] to avoid
/// unnecessary allocations.
final _compilationIdBuilder = VarintBuilder(32, 'compilation ID');

/// Parses a compilation ID and encoded protobuf message from a packet buffer as
/// specified in the embedded protocol.
(int, Uint8List) parsePacket(Uint8List packet) {
  try {
    var i = 0;
    while (true) {
      if (i == packet.length) {
        throw parseError(
            "Invalid compilation ID: continuation bit always set.");
      }

      var compilationId = _compilationIdBuilder.add(packet[i]);
      i++;
      if (compilationId != null) {
        return (compilationId, Uint8List.sublistView(packet, i));
      }
    }
  } finally {
    _compilationIdBuilder.reset();
  }
}
