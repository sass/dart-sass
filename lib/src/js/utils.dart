// Copyright 2017 Google Inc. Use of this source code is governed by an
// MIT-style license that can be found in the LICENSE file or at
// https://opensource.org/licenses/MIT.

import 'package:js_core/js_core.dart';
import 'package:node_interop/module.dart';
import 'package:node_interop/node_interop.dart';

import '../syntax.dart';
import '../value.dart';

/// Converts a JavaScript separator string into a [ListSeparator].
ListSeparator parseSeparator(String? separator) => switch (separator) {
      ' ' => ListSeparator.space,
      ',' => ListSeparator.comma,
      '/' => ListSeparator.slash,
      null => ListSeparator.undecided,
      _ => JSError.throwLikeJS(JSError('Unknown separator "$separator".')),
    };

/// Converts a syntax string to an instance of [Syntax].
Syntax parseSyntax(String? syntax) => switch (syntax) {
      null || 'scss' => Syntax.scss,
      'indented' => Syntax.sass,
      'css' => Syntax.css,
      _ => JSError.throwLikeJS(JSError('Unknown syntax "$syntax".')),
    };

/// The path to the Node.js entrypoint, if one can be located.
String? get entrypointFilename {
  if (requireNamespace.main?.filename case var filename?) {
    return filename;
  } else if (process.arguments case [_, String path, ...]) {
    return moduleModule.createRequireNamespace(path).resolve(path);
  } else {
    return null;
  }
}
