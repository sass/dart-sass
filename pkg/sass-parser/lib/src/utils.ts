// Copyright 2024 Google Inc. Use of this source code is governed by an
// MIT-style license that can be found in the LICENSE file or at
// https://opensource.org/licenses/MIT.

import * as postcss from 'postcss';

import {Node} from './node';

/**
 * A type that matches any constructor for {@link T}. From
 * https://www.typescriptlang.org/docs/handbook/mixins.html.
 */
// eslint-disable-next-line @typescript-eslint/no-explicit-any
export type Constructor<T> = new (...args: any[]) => T;

/**
 * An explicit field description passed to `cloneNode` that describes in detail
 * how to clone it.
 */
interface ExplicitClonableField<Name extends string> {
  /** The field's name. */
  name: Name;

  /**
   * Whether the field can be set to an explicit undefined value which means
   * something different than an absent field.
   */
  explicitUndefined?: boolean;
}

/** The type of field names that can be passed into `cloneNode`. */
type ClonableField<Name extends string> = Name | ExplicitClonableField<Name>;

/** Makes a {@link ClonableField} explicit. */
function parseClonableField<Name extends string>(
  field: ClonableField<Name>
): ExplicitClonableField<Name> {
  return typeof field === 'string' ? {name: field} : field;
}

/**
 * Creates a copy of {@link node} by passing all the properties in {@link
 * constructorFields} as an object to its constructor.
 *
 * If {@link overrides} is passed, it overrides any existing constructor field
 * values. It's also used to assign {@link assignedFields} after the cloned
 * object has been constructed.
 */
export function cloneNode<T extends Pick<postcss.Node, 'source'>>(
  node: T,
  overrides: Record<string, unknown> | undefined,
  constructorFields: ClonableField<keyof T & string>[],
  assignedFields?: ClonableField<keyof T & string>[]
): T {
  // We have to do these casts because the actual `...Prop` types that get
  // passed in and used for the constructor aren't actually subtypes of
  // `Partial<T>`. They use `never` types to ensure that various properties are
  // mutually exclusive, which is not compatible.
  const typedOverrides = overrides as Partial<T> | undefined;
  const constructorFn = node.constructor as new (defaults: Partial<T>) => T;

  const constructorParams: Partial<T> = {};
  for (const field of constructorFields) {
    const {name, explicitUndefined} = parseClonableField(field);
    let value: T[keyof T & string] | undefined;
    if (
      typedOverrides &&
      (explicitUndefined
        ? Object.hasOwn(typedOverrides, name)
        : typedOverrides[name] !== undefined)
    ) {
      value = typedOverrides[name];
    } else {
      value = maybeClone(node[name]);
    }
    if (value !== undefined) constructorParams[name] = value;
  }
  const cloned = new constructorFn(constructorParams);

  if (typedOverrides && assignedFields) {
    for (const field of assignedFields) {
      const {name, explicitUndefined} = parseClonableField(field);
      if (
        explicitUndefined
          ? Object.hasOwn(typedOverrides, name)
          : typedOverrides[name]
      ) {
        // This isn't actually guaranteed to be non-null, but TypeScript
        // (correctly) complains that we could be passing an undefined value to
        // a field that doesn't allow undefined. We don't have a good way of
        // forbidding that while still allowing users to override values that do
        // explicitly allow undefined, though.
        cloned[name] = typedOverrides[name]!;
      }
    }
  }

  cloned.source = node.source;
  return cloned;
}

/**
 * If {@link value} is a Sass node, a record, or an array, clones it and returns
 * the clone. Otherwise, returns it as-is.
 */
function maybeClone<T>(value: T): T {
  if (Array.isArray(value)) return value.map(maybeClone) as T;
  if (typeof value !== 'object' || value === null) return value;
  // The only records we care about are raws, which only contain primitives and
  // arrays of primitives, so structued cloning is safe.
  if (value.constructor === Object) return structuredClone(value);
  if (value instanceof postcss.Node) return value.clone() as T;
  return value;
}

/**
 * Converts {@link node} into a JSON-safe object, with the given {@link fields}
 * included.
 *
 * This always includes the `type`, `sassType`, `raws`, and `source` fields if
 * set. It converts multiple references to the same source input object into
 * indexes into a top-level list.
 */
export function toJSON<T extends Node>(
  node: T,
  fields: (keyof T & string)[],
  inputs?: Map<postcss.Input, number>
): object {
  // Only include the inputs field at the top level.
  const includeInputs = !inputs;
  inputs ??= new Map();
  let inputIndex = inputs.size;

  const result: Record<string, unknown> = {};
  if ('type' in node) result.type = (node as {type: string}).type;

  fields = ['sassType', 'raws', ...fields];
  for (const field of fields) {
    const value = node[field];
    if (value !== undefined) result[field] = toJsonField(field, value, inputs);
  }

  if (node.source) {
    let inputId = inputs.get(node.source.input);
    if (inputId === undefined) {
      inputId = inputIndex++;
      inputs.set(node.source.input, inputId);
    }

    result.source = {
      start: node.source.start,
      end: node.source.end,
      inputId,
    };
  }

  if (includeInputs) {
    result.inputs = [...inputs.keys()].map(input => input.toJSON());
  }
  return result;
}

/**
 * Converts a single field with name {@link field} and value {@link value} to a
 * JSON-safe object.
 *
 * The {@link inputs} map works the same as it does in {@link toJSON}.
 */
function toJsonField(
  field: string,
  value: unknown,
  inputs: Map<postcss.Input, number>
): unknown {
  if (typeof value !== 'object' || value === null) {
    return value;
  } else if (Array.isArray(value)) {
    return value.map((element, i) =>
      toJsonField(i.toString(), element, inputs)
    );
  } else if ('toJSON' in value) {
    if ('sassType' in value) {
      return (
        value as {
          toJSON: (field: string, inputs: Map<postcss.Input, number>) => object;
        }
      ).toJSON('', inputs);
    } else {
      return (value as {toJSON: (field: string) => object}).toJSON(field);
    }
  } else {
    return value;
  }
}

/**
 * Returns the longest string (of code units) that's an initial substring of
 * every string in
 * {@link strings}.
 */
export function longestCommonInitialSubstring(strings: string[]): string {
  let candidate: string | undefined;
  for (const string of strings) {
    if (candidate === undefined) {
      candidate = string;
    } else {
      for (let i = 0; i < candidate.length && i < string.length; i++) {
        if (candidate.charCodeAt(i) !== string.charCodeAt(i)) {
          candidate = candidate.substring(0, i);
          break;
        }
      }
      candidate = candidate.substring(
        0,
        Math.min(candidate.length, string.length)
      );
    }
  }
  return candidate ?? '';
}
