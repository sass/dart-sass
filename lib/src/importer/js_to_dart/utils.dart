// Copyright 2023 Google Inc. Use of this source code is governed by an
// MIT-style license that can be found in the LICENSE file or at
// https://opensource.org/licenses/MIT.

import 'package:node_interop/js.dart';

import '../../js/utils.dart';
import '../utils.dart';

/// Throws a JsError if [scheme] isn't a valid URL scheme.
void validateUrlScheme(String scheme) {
  if (!isValidUrlScheme(scheme)) {
    jsThrow(
        JsError('"$scheme" isn\'t a valid URL scheme (for example "file").'));
  }
}
