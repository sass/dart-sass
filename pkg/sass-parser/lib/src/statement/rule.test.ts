// Copyright 2024 Google Inc. Use of this source code is governed by an
// MIT-style license that can be found in the LICENSE file or at
// https://opensource.org/licenses/MIT.

import {GenericAtRule, Interpolation, Root, Rule, css, sass, scss} from '../..';
import * as utils from '../../../test/utils';

describe('a style rule', () => {
  let node: Rule;
  describe('with no children', () => {
    function describeNode(description: string, create: () => Rule): void {
      describe(description, () => {
        beforeEach(() => void (node = create()));

        it('has type rule', () => expect(node.type).toBe('rule'));

        it('has sassType rule', () => expect(node.sassType).toBe('rule'));

        it('has matching selectorInterpolation', () =>
          expect(node).toHaveInterpolation('selectorInterpolation', '.foo '));

        it('has matching selector', () => expect(node.selector).toBe('.foo '));

        it('has empty nodes', () => expect(node.nodes).toHaveLength(0));
      });
    }

    describeNode(
      'parsed as SCSS',
      () => scss.parse('.foo {}').nodes[0] as Rule
    );

    describeNode('parsed as CSS', () => css.parse('.foo {}').nodes[0] as Rule);

    describe('parsed as Sass', () => {
      beforeEach(() => {
        node = sass.parse('.foo').nodes[0] as Rule;
      });

      it('has matching selectorInterpolation', () =>
        expect(node).toHaveInterpolation('selectorInterpolation', '.foo\n'));

      it('has matching selector', () => expect(node.selector).toBe('.foo\n'));

      it('has empty nodes', () => expect(node.nodes).toHaveLength(0));
    });

    describe('constructed manually', () => {
      describeNode(
        'with an interpolation',
        () =>
          new Rule({
            selectorInterpolation: new Interpolation({nodes: ['.foo ']}),
          })
      );

      describeNode(
        'with a selector string',
        () => new Rule({selector: '.foo '})
      );
    });

    describe('constructed from ChildProps', () => {
      describeNode('with an interpolation', () =>
        utils.fromChildProps({
          selectorInterpolation: new Interpolation({nodes: ['.foo ']}),
        })
      );

      describeNode('with a selector string', () =>
        utils.fromChildProps({selector: '.foo '})
      );
    });
  });

  describe('with a child', () => {
    function describeNode(description: string, create: () => Rule): void {
      describe(description, () => {
        beforeEach(() => void (node = create()));

        it('has matching selectorInterpolation', () =>
          expect(node).toHaveInterpolation('selectorInterpolation', '.foo '));

        it('has matching selector', () => expect(node.selector).toBe('.foo '));

        it('has a child node', () => {
          expect(node.nodes).toHaveLength(1);
          expect(node.nodes[0]).toBeInstanceOf(GenericAtRule);
          expect(node.nodes[0]).toHaveProperty('name', 'bar');
        });
      });
    }

    describeNode(
      'parsed as SCSS',
      () => scss.parse('.foo {@bar}').nodes[0] as Rule
    );

    describeNode(
      'parsed as CSS',
      () => css.parse('.foo {@bar}').nodes[0] as Rule
    );

    describe('parsed as Sass', () => {
      beforeEach(() => {
        node = sass.parse('.foo\n  @bar').nodes[0] as Rule;
      });

      it('has matching selectorInterpolation', () =>
        expect(node).toHaveInterpolation('selectorInterpolation', '.foo\n'));

      it('has matching selector', () => expect(node.selector).toBe('.foo\n'));

      it('has a child node', () => {
        expect(node.nodes).toHaveLength(1);
        expect(node.nodes[0]).toBeInstanceOf(GenericAtRule);
        expect(node.nodes[0]).toHaveProperty('name', 'bar');
      });
    });

    describe('constructed manually', () => {
      describeNode(
        'with an interpolation',
        () =>
          new Rule({
            selectorInterpolation: new Interpolation({nodes: ['.foo ']}),
            nodes: [{name: 'bar'}],
          })
      );

      describeNode(
        'with a selector string',
        () => new Rule({selector: '.foo ', nodes: [{name: 'bar'}]})
      );
    });

    describe('constructed from ChildProps', () => {
      describeNode('with an interpolation', () =>
        utils.fromChildProps({
          selectorInterpolation: new Interpolation({nodes: ['.foo ']}),
          nodes: [{name: 'bar'}],
        })
      );

      describeNode('with a selector string', () =>
        utils.fromChildProps({selector: '.foo ', nodes: [{name: 'bar'}]})
      );
    });
  });

  describe('assigned a new selector', () => {
    beforeEach(() => {
      node = scss.parse('.foo {}').nodes[0] as Rule;
    });

    it("removes the old interpolation's parent", () => {
      const oldSelector = node.selectorInterpolation!;
      node.selectorInterpolation = '.bar';
      expect(oldSelector.parent).toBeUndefined();
    });

    it("assigns the new interpolation's parent", () => {
      const interpolation = new Interpolation({nodes: ['.bar']});
      node.selectorInterpolation = interpolation;
      expect(interpolation.parent).toBe(node);
    });

    it('assigns the interpolation explicitly', () => {
      const interpolation = new Interpolation({nodes: ['.bar']});
      node.selectorInterpolation = interpolation;
      expect(node.selectorInterpolation).toBe(interpolation);
    });

    it('assigns the interpolation as a string', () => {
      node.selectorInterpolation = '.bar';
      expect(node).toHaveInterpolation('selectorInterpolation', '.bar');
    });

    it('assigns the interpolation as selector', () => {
      node.selector = '.bar';
      expect(node).toHaveInterpolation('selectorInterpolation', '.bar');
    });
  });

  describe('stringifies', () => {
    describe('to SCSS', () => {
      describe('with default raws', () => {
        it('with no children', () =>
          expect(new Rule({selector: '.foo'}).toString()).toBe('.foo {}'));

        it('with a child', () =>
          expect(
            new Rule({selector: '.foo', nodes: [{selector: '.bar'}]}).toString()
          ).toBe('.foo {\n    .bar {}\n}'));
      });

      it('with between', () =>
        expect(
          new Rule({
            selector: '.foo',
            raws: {between: '/**/'},
          }).toString()
        ).toBe('.foo/**/{}'));

      describe('with after', () => {
        it('with no children', () =>
          expect(
            new Rule({selector: '.foo', raws: {after: '/**/'}}).toString()
          ).toBe('.foo {/**/}'));

        it('with a child', () =>
          expect(
            new Rule({
              selector: '.foo',
              nodes: [{selector: '.bar'}],
              raws: {after: '/**/'},
            }).toString()
          ).toBe('.foo {\n    .bar {}/**/}'));
      });

      it('with before', () =>
        expect(
          new Root({
            nodes: [new Rule({selector: '.foo', raws: {before: '/**/'}})],
          }).toString()
        ).toBe('/**/.foo {}'));
    });
  });

  describe('clone', () => {
    let original: Rule;
    beforeEach(() => {
      original = scss.parse('.foo {@bar}').nodes[0] as Rule;
      // TODO: remove this once raws are properly parsed
      original.raws.between = '  ';
    });

    describe('with no overrides', () => {
      let clone: Rule;
      beforeEach(() => {
        clone = original.clone();
      });

      describe('has the same properties:', () => {
        it('selectorInterpolation', () =>
          expect(clone).toHaveInterpolation('selectorInterpolation', '.foo '));

        it('selector', () => expect(clone.selector).toBe('.foo '));

        it('raws', () => expect(clone.raws).toEqual({between: '  '}));

        it('source', () => expect(clone.source).toBe(original.source));

        it('nodes', () => {
          expect(clone.nodes).toHaveLength(1);
          expect(clone.nodes[0]).toBeInstanceOf(GenericAtRule);
          expect(clone.nodes[0]).toHaveProperty('name', 'bar');
        });
      });

      describe('creates a new', () => {
        it('self', () => expect(clone).not.toBe(original));

        for (const attr of [
          'selectorInterpolation',
          'raws',
          'nodes',
        ] as const) {
          it(attr, () => expect(clone[attr]).not.toBe(original[attr]));
        }
      });

      describe('sets parent for', () => {
        it('nodes', () => expect(clone.nodes[0].parent).toBe(clone));
      });
    });

    describe('overrides', () => {
      describe('selector', () => {
        describe('defined', () => {
          let clone: Rule;
          beforeEach(() => {
            clone = original.clone({selector: 'qux'});
          });

          it('changes selector', () => expect(clone.selector).toBe('qux'));

          it('changes selectorInterpolation', () =>
            expect(clone).toHaveInterpolation('selectorInterpolation', 'qux'));
        });

        describe('undefined', () => {
          let clone: Rule;
          beforeEach(() => {
            clone = original.clone({selector: undefined});
          });

          it('preserves selector', () => expect(clone.selector).toBe('.foo '));

          it('preserves selectorInterpolation', () =>
            expect(clone).toHaveInterpolation(
              'selectorInterpolation',
              '.foo '
            ));
        });
      });

      describe('selectorInterpolation', () => {
        describe('defined', () => {
          let clone: Rule;
          beforeEach(() => {
            clone = original.clone({
              selectorInterpolation: new Interpolation({nodes: ['.baz']}),
            });
          });

          it('changes selector', () => expect(clone.selector).toBe('.baz'));

          it('changes selectorInterpolation', () =>
            expect(clone).toHaveInterpolation('selectorInterpolation', '.baz'));
        });

        describe('undefined', () => {
          let clone: Rule;
          beforeEach(() => {
            clone = original.clone({selectorInterpolation: undefined});
          });

          it('preserves selector', () => expect(clone.selector).toBe('.foo '));

          it('preserves selectorInterpolation', () =>
            expect(clone).toHaveInterpolation(
              'selectorInterpolation',
              '.foo '
            ));
        });
      });

      describe('raws', () => {
        it('defined', () =>
          expect(original.clone({raws: {before: '  '}}).raws).toEqual({
            before: '  ',
          }));

        it('undefined', () =>
          expect(original.clone({raws: undefined}).raws).toEqual({
            between: '  ',
          }));
      });
    });
  });

  describe('toJSON', () => {
    it('with empty children', () =>
      expect(scss.parse('.foo {}').nodes[0]).toMatchSnapshot());

    it('with a child', () =>
      expect(scss.parse('.foo {@bar}').nodes[0]).toMatchSnapshot());
  });
});
