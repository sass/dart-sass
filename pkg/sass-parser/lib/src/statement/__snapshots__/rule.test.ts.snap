// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`a style rule toJSON with a child 1`] = `
{
  "inputs": [
    {
      "css": ".foo {@bar}",
      "hasBOM": false,
      "id": "<input css _____>",
    },
  ],
  "nodes": [
    <@bar;>,
  ],
  "raws": {},
  "sassType": "rule",
  "selector": ".foo ",
  "selectorInterpolation": <.foo >,
  "source": <1:1-1:12 in 0>,
  "type": "rule",
}
`;

exports[`a style rule toJSON with empty children 1`] = `
{
  "inputs": [
    {
      "css": ".foo {}",
      "hasBOM": false,
      "id": "<input css _____>",
    },
  ],
  "nodes": [],
  "raws": {},
  "sassType": "rule",
  "selector": ".foo ",
  "selectorInterpolation": <.foo >,
  "source": <1:1-1:8 in 0>,
  "type": "rule",
}
`;
